docs/CIP_6_Pipelines_Registry.md

```diff
@@ -0,0 +1,309 @@
+Additionally, this means that users cannot create custom additional indices for their collections. For example, it may make sense
+that a user wants to not only control the parameters of the vector segment but also create a custom index for the metadata segment, create
+alternative vector segments with different parameters,
+
+6. Composition with Language Model Generation
+
+In the future, we may want to support the full RAG loop within Chroma. This means that we would want to support being able to take the output of retrieval and use it as input to a language model. This would require us to be able to compose the embedding function with a language model. This is not possible today, as the embedding function is not a first class citizen in Chroma.
+
+## Proposal
+
+We introduce three concepts in order to solve these problems - Pipelines, The Registry, and Storage
+
+### Pipelines
+
+A pipeline is a set of steps that are applied to the inputs to a chroma insert call. Broadly, everything including the segments written to are a part of the pipeline. The pipeline is responsible for loading the document, preprocessing the document, and writing the document to the segments. Pipeline steps are composable functions that can be chained together to form a pipeline. Each pipeline step is registered with a name, and can be referenced by that name. This registration can be thought of as similar to SQL functions.
+
+#### Pipeline Step Registration - The Registry
+
+For example, if you wanted to create a pipeline that chunked a document, you could do something like this:
+
+First, register The "Chunk" pipeline step with the name "Chunk" as follows:
+
+```python
+@pipeline_step(name="Chunk")
+def chunk(data: str) -> List[str]:
+    # Chunk the data
+    return chunked_data
+```
+ <!-- NOTE: ALTERNATIVE CONSIDERED = DO A CLASS BASED APPROACH, WHERE THE CLASS IS THE PIPELINE STEP -->
+
+Then you can create a pipeline that chunks a document as follows:
+```python
+to_add = Chunk(data)
+```
+
+Calling Chunk() does not actually execute the pipeline step, instead it returns a reference to the pipeline step. This reference is then passed to the add step, which is responsible for executing the pipeline step. This is done to allow for the pipeline step to be executed on the server-side. A pipeline can always be executed locally by calling the run() method on the pipeline.
+
+The assumption then, is that you provide the server with the relevant pipeline steps to execute. The exact mechanism for this is later defined. For local execution, the pipeline steps are executed on the client-side.
+
+#### Pipeline Step Composition
+
+Pipeline steps can be composed, and return a reference to the composed pipeline step. For example, if you wanted to create a pipeline that chunked a document and then embedded the chunks, you could do something like this:
+
+```python
+to_add = Embed(Chunk(data))
```

Comment:
how can we make it easy to see what things can be chained? can we use type hints?

Reply:
yes.

Reply:
@HammadB, I love functional programming as much as the next guy, but do you think the same result can be achieved with a fluid (chained) API?

Reply:
@tazarov can you provide an example?

Reply:
An "extra" late-night thought - wdyt of YAML-based manifest, as a portable version of this (python functions must be bundled or can be installed as deps)

Reply:
@jeffchuber 

`withData(data).pipe([Embed,Chunk])`

or even

`with(data).pipe(Embed).pipe(Chunk)`

or even

`data | Embed | Chunk` (with operator overloading this is very pythonic)

Reply:
Yeah I'm not wedded to this specific interface tbh

Reply:
i still lean towards "its just a function" because I think its the easiest to understand and more importantly -- I think it may be the easiest to "break out of" when you want to add a new operation.

Reply:
I like "withData(data).pipe([Embed,Chunk])" as the function approach could look messy once you pass 5 pipelines, which is feasible at the moment, : FixMetadata, AssignIDs, ParentChunking, Chunking, Embedding, etc

Reply:
Using `|` operator really conveys the pipe in the pipeline, and for most devs (familiar with bash), pipes will look natural and also Pythonic. Another aspect is if one wants to pass parameters to steps, it gets even messier.

Another aspect of using the chained method calls is that we can even use Lambdas for filtering and inline processing without need for decorators:

```python
withData(data).pipe([lambda x: 'allowed' in x,Chunk,Embed])
```

Reply:
Came here to suggest a fluid/chained API, but i'm quite late to the party :)

Total subjective IMO, but i've tended to rely on using indentation when wrapping lots of functions together ([clojure transducers](https://practical.li/blog/posts/transducers-in-clojure-getting-started/) jump to mind) as a way of visually communicating the level that code is at.

As a python noob, i genuinely don't know, but i wonder if such indentation tricks are  possible with chaining in a way they're not with function wrapping. Like, i believe this is allowable, where a particular indentation level clearly indicates where each step definition starts:

```python
(
    withData(data)
        .pipe(Chunk)
        .pipe(Embed)
)
```

But idk if there's an equivalently clear way of doing it with function wrapping?